include/deal.II/base/numbers.h

```diff
@@ -192,6 +192,29 @@ namespace numbers
    */
   bool is_finite (const std::complex<long double> &x);
 
+  /**
+   * Return whether two numbers are equal to one another. For intricate data
+   * types (e.g. some automatically differentiable numbers), this function
+   * returns only whether the scalar values stored by the input values are
+   * equal.
+   *
+   * @note This function expects that @p value_2 is castable to the type
+   * of @p value_1.
+   */
+  template <typename Number1,typename Number2>
+  bool
+  are_equal_valued (const Number1 &value_1, const Number2 &value_2);
```

Comment:
You can enforce the requirement stated in the documentation above in the function signature if you write it as
```
template <typename Number>
bool
￼ +  are_equal_valued (const Number &value_1, const typename identity<Number>::type &value_2);
```
This way, the compiler will infer `Number` from the first argument, and the second argument will be cast to the same type in the call already. The call will not compile if the second argument cannot be cast to the first argument's type.